crypto/x509/x509_vfy.c

```diff
@@ -3127,7 +3139,7 @@ static int build_chain(X509_STORE_CTX *ctx)
                         search = 0;
                         continue;
                     }
-                    if (ss == 0)
+                    if (!self_issued)
```

Comment:
This check in the code is what the adapted comment in (new) line 3116 refers to.
So the explanation for the change is what I wrote above.

If we were to allow continuing with the chain building despite the fact that the issuer of the certs just added is the same as its subject (which is possible in theory and appears not excluded by RFC 5280, in particular https://tools.ietf.org/html/rfc5280#section-6.1) we would need to be careful not to include the same cert again. 
(Such a generalization AFAICS could make sense but is not intended in this PR.)

Reply:
It seems you've not looked at the current code closely enough.
* We do handle self-issued certificates without looping, when we look for a new issuer to add to the chain, we specifically exclude adding certificates that are already in the chain.
* We don't treat certificates where the akid doesn't match as self-signed.  Self-issued certificates should have indicate the real signer via the akid.
* The trust anchors are self-signed. If we've not built the chain all the way to the trust-anchor it is incomplete and not signed by the trust-anchor key.

Barring strong evidence, this change is wrong, and needs to be taken out.

Reply:
Correct that I do not have the full picture of the `build_chain()` function, so indeed the conclusions I took from looking at the code section (including the comment discussed above) can be wrong. Interestingly, the experiments/tests I did a week ago lead me to the conclusion and change stated above but for some reason I cannot reproduce that any more - at least the current respective module tests: `make test TESTS="test_verify test_x509"` pass both with using `!self_signed` and with using instead `!self_issued`.

I'm currently confused on this outcome, and given your most recent comment I'm going to revert this part of the PR soon (unless a strong indication is brought up in favor of the current variant, as you wrote).

Reply:
Is there a detailed documentation of `build_chain()` or of a more high-level/public variant of it?
The only - pretty superficial - one I found is in `doc/man3/SSL_CTX_add1_chain_cert.pod`: 
```
SSL_CTX_build_cert_chain() builds the certificate chain for B<ctx> normally
this uses the chain store or the verify store if the chain store is not set.
If the function is successful the built chain will replace any existing chain.
The B<flags> parameter can be set to B<SSL_BUILD_CHAIN_FLAG_UNTRUSTED> to use
existing chain certificates as untrusted CAs, B<SSL_BUILD_CHAIN_FLAG_NO_ROOT>
to omit the root CA from the built chain, B<SSL_BUILD_CHAIN_FLAG_CHECK> to
use all existing chain certificates only to build the chain (effectively
sanity checking and rearranging them if necessary), the flag
B<SSL_BUILD_CHAIN_FLAG_IGNORE_ERROR> ignores any errors during verification:
if flag B<SSL_BUILD_CHAIN_FLAG_CLEAR_ERROR> is also set verification errors
are cleared from the error queue.
```

Reply:
> Is there a detailed documentation of `build_chain()` or of a more high-level/public variant of it?

The `build_chain()` function is a static (local) helper function in `crypto/x509/x509_vfy.c` so there's naturally no documentation for that.  It is however the "bottom half" of `X509_verify_cert(3)`, which verifies a chain associated with an `X509_STORE_CTX` that is configured per `X509_STORE_CTX_init(3)` with verification parameters that are configurable via `X509_STORE_CTX_get0_param(3)` or `X509_STORE_CTX_set0_param(3)`, which leads us to the batch of functions grouped with `X509_VERIFY_PARAM_set_flags(3)`.

The `SSL_CTX_build_cert_chain(3)` tries to construct a more complete chain for the local (typically server) certificate, probably not the droid you're looking for.

`X509_verify_cert(3)` implements two trust models, WebPKI and DANE.  Since DANE is all about TLS (DNS records for TLS services at a given port/protocol/host) it is documented on the SSL side of the interface under `SSL_CTX_dane_enable(3)`, but there of course needs to be a "bridge" between `libssl` and `libcrypto` to expose the DANE settings to `X509_verify_cert(3)`. That bridge is currently an internal interface (not documented), but it is an exposed `libcrypto` function: `X509_STORE_CTX_set0_dane`.

We can imagine some creative use of that interface outside `libssl` and the test-suite. Someone might want to write a command-line DANE validator (somewhat along the lines of `test/danetest.c`) that takes a cert chain and a set of TLSA records, and performs the requisite validation. The TLSA records are configured for an `SSL` handle that is initialized a la DANE, but not actually used to connect with any peer. Instead the handle is milked for its DANE settings to apply to the `X509_STORE_CTX`:
```
X509_STORE_CTX_set0_dane(store_ctx, SSL_get0_dane(ssl));
```
at which point, presto magic, you can perform DANE verification of the chain for that `store_ctx`.

Bottom line, when chains are constructed for verifying the EE and additional untrusted certificates provided by the peer, the chain is constructed based on various X509_VERIFY_PARAM settings, the associated trust store and its settings, any DANE parameters, and verification callbacks.

Minimal parameters are initially configured via `X509_STORE_CTX_set_default(3)`, with additional parameters modifying these (e.g. the "purpose" is often inherited from the default context).

The trusted certificate store can be replaced with a custom stack of trusted CAs via `X509_STORE_CTX_set0_trusted_stack(3)`, while `X509_STORE_CTX_set0_untrusted(3)`, which just gets/sets the peer's untrusted cert stack, was needed for my out of tree DANE library in OpenSSL 1.0.0--1.0.2, which is still used in Postfix, Exim, ... until all DANE users stop support for 1.0.x, and switch to the 1.1.x DANE API.

I hope that's helpful. The X509 library has many moving parts.  The `build_chain()` function replaced up a rather fragile predecessor (with a history of bugs), but the overall division of labour between the various parts of the crypto library has not changed. It takes some considerable effort to read enough of it to get a good sense of what's going on. There are stores, purposes, verify parameters, regular and trusted certificates, policies (which almost nobody uses), ad-hoc saner policies on intermediate certs via EKUs, ...

If you have more questions, feel free to ask...

Reply:
Thanks a lot @vdukhovni for your explanations. Yet they only partially address the point here, which is not DANE neither cert path validation, but details on X.509 cert path _construction_.

Clear that `build_chain()` is internal, but this does not by itself exclude documentation (e.g., in `doc/internal/man3`). Actually I do not only miss documentation for this important and interesting (but of course not security-critical) aspect of `X509_verify_cert()`: what really is missing is a proper/official way of using the chain construction for purposes such as in TLS and CMP (and certainly many more) where a party is supposed to send along with its own (e.g., data signer) cert the chain of this cert (excluding the root), usually called 'extra certs' or the like. 

During the ongoing CMP contribution I've brought in a function (along with internal documentation) that does this job: `ossl_cmp_build_cert_chain()` in `crypto/cmp/cmp_util.c`.
Yet this function unfortunately actually does too much: by using `X509_verify_cert()`it also attempts to verify the chain because it cannot call `build_chain()` directly, which results in a waste of computational resources and also requires chopping off any new error queue entries. 

I'm going to file an issue to add and document and OpenSSL API function for cert path/chain construction. Its documentation should state the input needed (i.e., target cert, untrusted certs, maybe trusted certs) and the effect of applicable verification flags such as 'partial chain'.

What I was actually after for solving the current issue here was some detail on how the actual chain building works: which certs are considered when and how the chaining is done. Your response already contains some hints in this direction, namely that the essential thing is the AKID and not the issuer DN, and that there is already a check that aims at preventing loops.

Reply:
My understanding is that the historical view has been that if you have a certificate, you got a chain along with it, from the same entity (CA) that gave you the certificate itself.  So, the two are essentially bundled together as configuration and you wouldn't need to build a chain before you send your certificate, since the chain is just configured along with the certificate itself.  I guess the CMP is changing that assumption?

Reply:
CMP does not change anything about the use of cert chains and self-issued/self-signed certs.
The adaptations to cert chain building and cert verification in this PR have little to do with CMP, except for the side topic I've mentioned above:
When the first CMP message (similarly to the first TLS package) is signature-protected the sender should include in the extraCerts the chain of the protecting cert.
Maybe the sender has kept record of this chain as you wrote, but in general the chain needs to be re-constructed (both in CMP and TLS implementations) from any set of 'untrusted' candidate intermediate CA certs available to the sender. To this end it would be nice if the internal chain building function was not only better documented but officially usable not only by the TLS implementation but by any other modules/applications (as it is the case for other crypto libs).

Reply:
Actually CMP has also a second use case where a (rather) direct call to `build_chain()` would be helpful: when the CA/server returns a newly enrolled cert to the client and it protects the response message with a signature, the chain of the sender cert is in general different to the chain of the newly issued cert. Due to a design weakness of CMP the server typically appends/mixes the two chains when sending them in the extraCerts field of the response message, and the client then has the ugly job to find out which of the extraCerts belong to which chain.

Reply:
@vdukhovni, I meanwhile have been able to rediscover the situation where `build_chain()` starts looping unless `!self_signed` is replaced by `!self_issued`. Yet this turned out to happen only after I had introduced a mistake in `X509_check_issued()`:
```
    if issuer == subject) /* self-issued */
        return X509_V_OK;
```
which I meanwhile corrected.
So as discussed above I've returned to (essentially) the original condition `(ss == 0)`, 
which after refactoring (for better readability etc.) looks like this: `(!self_signed)`.

Reply:
You likely already know this, but there is already a function for augmenting (building) a chain for one's own certificate in `libssl`, that is `ssl_build_cert_chain()` in `ssl/ssl_cert.c`. It calls `X509_verify_cert()` for its side-effect of chain construction.  It is exposed via the `SSL_CTX_build_cert_chain()` and `SSL_build_cert_chain()` control macros.

I'm not fond of configurations where the desired chain is not explicitly specified, and don't use this interface, but it is available to applications that want to use it.

Reply:
Doing some further improvements on the descriptions in `doc/man1/openssl-verify.pod.in` I realized that that this doc file does include a pretty good documentation of the activity of the internal `build_chain()` function. The respective text passages under the `VERIFY OPERATION` header should be well usable as documentation of the API function I've requested in #10696.